HPCC-17737 Config Manager should handle new comp type requests

rpastrana opened the pull request:
- Adds new front-end "Computer Type" request type
- Supports new "Computer Type" requests in WsDeploy

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana committed:
HPCC-17737 Config Manager should handle new comp type requests

- Adds new comptype request type
- Supports new comptype request types

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@wangkx please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 93fbc8f3f9612b46ca25c2c9a2ac8aaa6974a915
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.284 seconds |


Install hpccsystems-platform-community_6.4.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |730 |730 |0 |
|test (thor) |624 |624 |0 |
|test (roxie) |758 |758 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

wangkx commented on esp/files/scripts/configmgr/configmgr.js:
```diff
@@ -3312,27 +3313,30 @@ function onMenuItemClickHandleComputer(p_sType, p_aArgs, p_oValue) {
+  {
+    if (parentName === "Computers")
+    {
+      for (i = 0; i < recSet.getLength(); i++) {
```
Move '{' to the next line for consistence?

wangkx commented on esp/files/scripts/configmgr/configmgr.js:
```diff
@@ -3312,27 +3313,30 @@ function onMenuItemClickHandleComputer(p_sType, p_aArgs, p_oValue) {
-      if (YAHOO.lang.isArray(arr))
-        break;
-    }
+      if (i >= recSet.length) {
```
Same to my last comment.

wangkx commented on esp/files/scripts/configmgr/configmgr.js:
```diff
@@ -3312,27 +3313,30 @@ function onMenuItemClickHandleComputer(p_sType, p_aArgs, p_oValue) {
-    else
-      top.document.navDT.promptNewRange([], [], type);
-    return;
+    else if (parentName ==="Computer Types")
```
Change 'else if' to 'if'?

rpastrana replied:
It might be difficult to see due to existing indentation, but this is meant to pair up w/ the " if (parentName === "Computers")" of line 3317.
However, I prob should consider the unlikely "else" case...

wangkx replied:
There is a 'return' for  if (parentName === "Computers")" { ... }.

wangkx commented on esp/services/WsDeploy/WsDeployService.cpp:
```diff
@@ -4811,6 +4810,25 @@ bool CWsDeployFileInfo::handleComputer(IEspContext &context, IEspHandleComputerR
+      StringBuffer sbNewName(type);
+      xpath.clear().appendf("%s", type);
+
+      getUniqueName(pEnvRoot, sbNewName, xpath.str(), XML_TAG_HARDWARE);
```
Replace 'xpath.str()' with 'type'?

wangkx reviewed (commented):
a few minor comments.

rpastrana commented:
@richardkchapman this is a fix to a regression I introduced. 
The reason the regression was inadvertently introduced has to do with the original code handling multiple request types ("New Computer Type" and "New Computer (Instance)") via a common switch case path. The front end was bundling both types into a generic named "NEW" request.
Admittedly, this regression should have been caught during system tests (if non-related actions would have been tested), or during development if I could have realized that a single code path was handling multiple request types. 
Going forward, configmgr code changes should be analyzed for similar behavior. FYI @kenrowland

rpastrana committed:
HPCC-17737 Config Manager should handle new comp type requests

- Handles Conputer Type, Switches, Domains, NAS as computer Element requests
- Addresses Curly bracket format
- Renames "New" requests to "NewComputer" requests
- Adds restriction on prefix value when creating new comp or range
- Removes unnecessary code

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@wangkx please review latest changes.

wangkx commented:
@rpastrana looks fine.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: b85f3a066e827a54c56597656c080dc8723a1da7
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 93.866 seconds |


Install hpccsystems-platform-community_6.4.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |730 |730 |0 |
|test (thor) |624 |624 |0 |
|test (roxie) |758 |758 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented:
@richardkchapman please merge.
